pkg/cloud/services/securitygroup/securitygroups.go

```diff
@@ -363,6 +368,55 @@ func (s *Service) describeSecurityGroupsByName() (map[string]infrav1.SecurityGro
 	return res, nil
 }
 
+// revokeIngressAndEgressRulesFromVPCDefaultSecurityGroup revokes ingress and egress rules from the VPC default security group.
+// The VPC default security group is created by AWS and cannot be deleted.
+// But we can revoke all ingress and egress rules from it to make it more secure. This security group is not used by CAPA.
+func (s *Service) revokeIngressAndEgressRulesFromVPCDefaultSecurityGroup() error {
+	if !s.scope.VPC().EmptyRoutesDefaultVPCSecurityGroup {
+		return nil
+	}
+
+	securityGroups, err := s.EC2Client.DescribeSecurityGroupsWithContext(context.TODO(), &ec2.DescribeSecurityGroupsInput{
+		Filters: []*ec2.Filter{
+			filter.EC2.VPC(s.scope.VPC().ID),
+			filter.EC2.SecurityGroupName("default"),
+		},
+	})
+	if err != nil {
+		return errors.Wrapf(err, "failed to find default security group in vpc %q", s.scope.VPC().ID)
+	}
+	defaultSecurityGroupID := *securityGroups.SecurityGroups[0].GroupId
+	s.scope.Debug("Removing ingress and egress rules from default security group in VPC", "defaultSecurityGroupID", defaultSecurityGroupID, "vpc-id", s.scope.VPC().ID)
+
+	ingressRules := infrav1.IngressRules{
+		{
+			Protocol:               infrav1.SecurityGroupProtocolAll,
+			FromPort:               -1,
+			ToPort:                 -1,
+			SourceSecurityGroupIDs: []string{defaultSecurityGroupID},
+		},
+	}
+	err = s.revokeSecurityGroupIngressRules(defaultSecurityGroupID, ingressRules)
+	if awserrors.IsIgnorableSecurityGroupError(errors.Cause(err)) != nil {
```

Comment:
Are `GroupNotFound`/`PermissionNotFound` (which `IsIgnorableSecurityGroupError` handles) really errors that we want to ignore? The default group cannot be deleted, so such errors here should fail, I think. From the test, it seems that you found that AWS throws a permission error if the given rules do not exist. So we could instead use `ec2:DescribeSecurityGroupRules` (new permission which CAPA does not have yet) or handle the very specific error?

Reply:
After deleting the ingress/egress rules from the default security group, the next time we try to delete them we'll get a `PermissionNotFound` error. I think this one can be ignored, because that's the state we want. To avoid swallowing the `GroupNotFound` error, that should not happen in this scenario, we could handle `PermissionNotFound` more specifically, like you mentioned. I'll add a new `IsPermissionNotFoundError` function. How does that sound?